File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelRunner.java
package org.jetbrains.bsp.bazel.bazelrunner;

import com.google.common.collect.Lists;
import io.vavr.control.Option;
import java.io.IOException;
import java.util.List;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.bsp.bazel.logger.BspClientLogger;

public class BazelRunner {

  private static final Logger LOGGER = LogManager.getLogger(BazelRunner.class);

  private static final String PUBLISH_ALL_ACTIONS = "--build_event_publish_all_actions";
  private static final String BES_BACKEND = "--bes_backend=grpc://localhost:";

  private final BazelPathProvider bazelPathProvider;

  private Option<Integer> besBackendPort = Option.none();
  private final BspClientLogger bspClientLogger;
  private final BazelInfo bazelInfo;

  private final List<String> defaultFlags;

  // This is runner without workspace path. It is used to determine workspace
  // path and create a fully functional runner.
  public static BazelRunner inCwd(BazelPathProvider bazelPath, BspClientLogger bspClientLogger) {
    return new BazelRunner(bazelPath, bspClientLogger, null, List.of());
  }

  public static BazelRunner of(
      BazelPathProvider bazelPath,
      BspClientLogger bspClientLogger,
      BazelInfo bazelInfo,
      List<String> defaultFlags) {
    return new BazelRunner(bazelPath, bspClientLogger, bazelInfo, defaultFlags);
  }

  private BazelRunner(
      BazelPathProvider bazelPathProvider,
      BspClientLogger bspClientLogger,
      BazelInfo bazelInfo,
      List<String> defaultFlags) {
    this.bazelPathProvider = bazelPathProvider;
    this.bspClientLogger = bspClientLogger;
    this.bazelInfo = bazelInfo;
    this.defaultFlags = defaultFlags;
  }

  public BazelRunnerCommandBuilder commandBuilder() {
    return new BazelRunnerCommandBuilder(this);
  }

  BazelProcess runBazelCommandBes(String command, List<String> flags, List<String> arguments) {
    List<String> newFlags = getBesFlags(flags);
    return runBazelCommand(command, newFlags, arguments);
  }

  private List<String> getBesFlags(List<String> flags) {
    List<String> newFlags = Lists.newArrayList(getBesBackendAddress(), PUBLISH_ALL_ACTIONS);
    newFlags.addAll(flags);

    return newFlags;
  }

  private String getBesBackendAddress() {
    var port = besBackendPort.getOrElseThrow(() -> new IllegalStateException("BES port not set"));
    return BES_BACKEND + port;
  }

  BazelProcess runBazelCommand(String command, List<String> flags, List<String> arguments) {
    try {
      return runBazelProcess(command, flags, arguments);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private synchronized BazelProcess runBazelProcess(
      String command, List<String> flags, List<String> arguments) throws IOException {
    var processArgs = getProcessArgs(command, flags, arguments);

    var message = "Invoking: " + String.join(" ", processArgs);
    LOGGER.info(message);
    bspClientLogger.message(message);

    var processBuilder = new ProcessBuilder(processArgs);
    if (bazelInfo != null) {
      processBuilder.directory(bazelInfo.workspaceRoot().toFile());
    }
    var process = processBuilder.start();

    return new BazelProcess(process, bspClientLogger);
  }

  private List<String> getProcessArgs(String command, List<String> flags, List<String> arguments) {
    var processArgs = Lists.newArrayList(bazelPathProvider.currentBazelPath(), command);
    processArgs.addAll(defaultFlags);
    processArgs.addAll(flags);
    processArgs.addAll(arguments);
    return processArgs;
  }

  public void setBesBackendPort(int port) {
    besBackendPort = Option.of(port);
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/BazelRunner.kt
package org.jetbrains.bsp.bazel.bazelrunner

import org.apache.logging.log4j.LogManager
import org.jetbrains.bsp.bazel.bazelrunner.params.BazelFlag.repositoryOverride
import org.jetbrains.bsp.bazel.commons.Constants
import org.jetbrains.bsp.bazel.logger.BspClientLogger
import org.jetbrains.bsp.bazel.workspacecontext.WorkspaceContext
import org.jetbrains.bsp.bazel.workspacecontext.WorkspaceContextProvider
import org.jetbrains.bsp.bazel.workspacecontext.extraFlags
import java.nio.file.Path

class BazelRunner private constructor(
    private val workspaceContextProvider: WorkspaceContextProvider,
    private val bspClientLogger: BspClientLogger,
    val workspaceRoot: Path?,
    val bazelBspRoot: String
) {

    companion object {
        private val LOGGER = LogManager.getLogger(BazelRunner::class.java)

        @JvmStatic
        fun of(
            workspaceContextProvider: WorkspaceContextProvider,
            bspClientLogger: BspClientLogger,
            workspaceRoot: Path?,
            bazelBspRoot: String
        ): BazelRunner {
            return BazelRunner(workspaceContextProvider, bspClientLogger, workspaceRoot, bazelBspRoot)
        }
    }

    fun commandBuilder(): BazelRunnerCommandBuilder = BazelRunnerCommandBuilder(this)

    fun runBazelCommandBes(
        command: List<String>,
        flags: List<String>,
        arguments: List<String>,
        environment: Map<String, String>,
        originId: String?,
        eventTextFile: Path,
        serverPid: Long?,
    ): BazelProcess {
        fun besFlags() = listOf(
            "--build_event_binary_file=${eventTextFile.toAbsolutePath()}",
            "--bes_outerr_buffer_size=10",
            "--isatty=true",
        )

        // TODO https://youtrack.jetbrains.com/issue/BAZEL-617
        return runBazelCommand(
            command,
            flags = besFlags() + flags,
            arguments,
            environment,
            originId,
            true,
        )
    }

    fun runBazelCommand(
        command: List<String>,
        flags: List<String>,
        arguments: List<String>,
        environment: Map<String, String>,
        originId: String?,
        parseProcessOutput: Boolean,
        useBuildFlags: Boolean = true,
        needsServerPid: Boolean = true,
    ): BazelProcess {
        val workspaceContext = workspaceContextProvider.currentWorkspaceContext()
        val usedBuildFlags = if (useBuildFlags) buildFlags(workspaceContext) else emptyList()
        val defaultFlags = listOf(repositoryOverride(Constants.ASPECT_REPOSITORY, bazelBspRoot))
        val processArgs =
            listOf(bazel(workspaceContext)) + command + usedBuildFlags + defaultFlags + flags + arguments
        logInvocation(processArgs, environment, originId)
        val processBuilder = ProcessBuilder(processArgs)
        processBuilder.environment() += environment
        val outputLogger = bspClientLogger.takeIf { parseProcessOutput }?.copy(originId = originId)
        workspaceRoot?.let { processBuilder.directory(it.toFile()) }
        val serverPid = if (needsServerPid) resolveBazelServerPid() else null
        val process = processBuilder.start()
        return BazelProcess(
            process,
            outputLogger,
            serverPid,
        )
    }

    private fun envToString(environment: Map<String, String>): String =
        environment.entries.joinToString(" ") { "${it.key}=${it.value}" }

    private fun logInvocation(processArgs: List<String>, processEnv: Map<String, String>, originId: String?) {
        "Invoking: ${envToString(processEnv)} ${processArgs.joinToString(" ")}"
            .also { LOGGER.info(it) }
            .also { bspClientLogger.copy(originId = originId).message(it) }
    }

    private fun bazel(workspaceContext: WorkspaceContext): String = workspaceContext.bazelBinary.value.toString()
    private fun buildFlags(workspaceContext: WorkspaceContext): List<String> =
        workspaceContext.buildFlags.values + workspaceContext.extraFlags

    private fun resolveBazelServerPid(): Long? {
        val processResult = commandBuilder()
            .info()
            .withArgument("server_pid")
            .executeBazelCommand(needsServerPid = false)
            .waitAndGetResultTimeout()
        return processResult.stdoutLines.firstOrNull()?.toLong()
    }
}
